alica_ros_proxy/include/clock/AlicaRosTimer.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+#include "engine/IAlicaTimer.h"
+#include "engine/AlicaClock.h"
+
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <ros/callback_queue.h>
+#include <ros/callback_queue_interface.h>
+#include <ros/ros.h>
+
+namespace alicaRosTimer
+{
+
+template <class CallbackQ>
+class SyncStopTimerRosImpl : public std::enable_shared_from_this<SyncStopTimerRosImpl<CallbackQ>>
+{
+    using Base = std::enable_shared_from_this<SyncStopTimerRosImpl<CallbackQ>>;
+public:
+    using TimerCb = std::function<void()>;
+
+    SyncStopTimerRosImpl(CallbackQ& cbQ, TimerCb&& userCb, alica::AlicaTime period)
+            : _userCb(std::move(userCb))
+            , _period(toRosDuration(period))
+            , _nh()
+            , _timer()
+            , _stopMutex()
+            , _stopCv()
+            , _userCbInProgress(false)
+            , _stop(false)
+    {
+        _nh.setCallbackQueue(std::addressof(cbQ));
+    }
+
+    void start()
+    {
+        // Create a one shot timer to fire the first callback immediately. One shot timer is needed because ROS timers don't provide a
+        // a way to create a normal timer whose first callback is immediate
+        // Grab a weak ptr to this object (grabbing a shared_ptr will result in a cycle)
+        _timer = _nh.createTimer(ros::Duration(0), [weak_ptr_self = Base::weak_from_this()](auto&&... args) {
+            // Ensure impl object for this timer is not destroyed
+            if (auto self = weak_ptr_self.lock()) {
+                self->timerCb();
+            }
+        }, true, true);
+    }
```

Comment:
above this line

            // Schedule the next timer event here so that multiple timerCb's are not active at the same time
            _timer.setPeriod(_period);
            // Need to stop & then start the timer for it to fire again, just a start call will be ignored by ROS
            _timer.stop();
            _timer.start();

Reply:
From going through the ros docs it does not seem like a oneshot timer can be converted to a normal timer. It can only be made to fire again by calling setPeriod, stop & then start so we have to do this inside the callback